encoding/gzip/gzip.go

```diff
@@ -0,0 +1,79 @@
+/*
+ *
+ * Copyright 2017 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package gzip
+
+import (
+	"compress/gzip"
+	"io"
+	"io/ioutil"
+	"sync"
+
+	"google.golang.org/grpc/encoding"
+)
+
+// New function registers gzip compressor. It maintains two pools for compressor and decompressor.
+func New() encoding.Compressor {
+	return &compressor{
+		name: "gzip_register",
+		poolCompressor: sync.Pool{
+			New: func() interface{} {
+				return gzip.NewWriter(ioutil.Discard)
+			},
+		},
+		poolDecompressor: sync.Pool{},
+	}
+}
+
+func (c *compressor) Compress(w io.Writer) (io.WriteCloser, error) {
+	z := c.poolCompressor.Get().(*gzip.Writer)
```

Comment:
We never `Put()` the compressor.

I think this needs to be a little more complicated.  We need to override the `Close()` method on the WriteCloser returned by `gzip.NewWriter` so that it calls `poolCompressor.Put()` on itself.  LMK if you need help with this.